vortex-cxx/src/lib.rs

```diff
@@ -0,0 +1,339 @@
+
+        type VortexScanBuilder;
+        fn scan_builder_set_filter(builder: &mut VortexScanBuilder, filter: &[u8]) -> Result<()>;
+        fn scan_builder_set_limit(builder: &mut VortexScanBuilder, limit: usize);
+        unsafe fn scan_builder_into_arrow(
+            builder: Box<VortexScanBuilder>,
+            out_array: *mut u8,
+            out_schema: *mut u8,
+        ) -> Result<()>;
+        unsafe fn scan_builder_into_stream(
+            builder: Box<VortexScanBuilder>,
+            out_stream: *mut u8,
+        ) -> Result<()>;
+
+        type VortexWriteOptions;
+        fn write_options_new() -> Box<VortexWriteOptions>;
+        unsafe fn write_array_stream(
+            options: Box<VortexWriteOptions>,
+            input_stream: *mut u8,
+            path: &str,
+        ) -> Result<()>;
+
+        fn configure_runtime(worker_threads: usize) -> Result<()>;
+    }
+}
+
+pub struct VortexFile {
+    inner: vortex::file::VortexFile,
+}
+
+/// File operations - using blocking operations for simplicity
+/// TODO(xinyu): object store (see vortex-ffi)
+fn open_file(path: &str) -> Result<Box<VortexFile>, Box<dyn std::error::Error + Send + Sync>> {
+    let file = VortexOpenOptions::file().open_blocking(std::path::Path::new(path))?;
+    Ok(Box::new(VortexFile { inner: file }))
+}
+
+fn file_row_count(file: &VortexFile) -> u64 {
+    file.inner.row_count()
+}
+
+fn file_scan_builder(
+    file: &VortexFile,
+) -> Result<Box<VortexScanBuilder>, Box<dyn std::error::Error + Send + Sync>> {
+    Ok(Box::new(VortexScanBuilder {
+        inner: file.inner.scan()?,
+    }))
+}
+
+pub struct VortexScanBuilder {
+    inner: ScanBuilder<ArrayRef>,
+}
+
+fn scan_builder_set_filter(
+    builder: &mut VortexScanBuilder,
+    filter: &[u8],
+) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
+    let filter = deserialize_expr_proto(&Expr::decode(filter)?, &ExprRegistry::default())
+        .map_err(|e| e.with_context("deserializing filter expr"))?;
+    take_mut::take(&mut builder.inner, |inner| inner.with_filter(filter));
+    Ok(())
+}
+
+fn scan_builder_set_limit(builder: &mut VortexScanBuilder, limit: usize) {
+    // Overwrite inner without dropping it.
+    take_mut::take(&mut builder.inner, |inner| inner.with_limit(limit));
+}
+
+/// Arrow Rust FFI interplay with C++ best practices refer to:
+/// https://github.com/dora-rs/dora/blob/42775e3612b34d3998b1f7feb7e5df7ec3f8a7bd/examples/c%2B%2B-arrow-dataflow/node-rust-api/main.cc#L12-L19
+/// https://github.com/dora-rs/dora/blob/42775e3612b34d3998b1f7feb7e5df7ec3f8a7bd/apis/c%2B%2B/node/src/lib.rs#L211-L212
+///
+/// # Safety
+///
+/// out_array should be properly aligned according to C ABI and valid for write.
+/// out_schema should be properly aligned according to C ABI and valid for write.
+unsafe fn scan_builder_into_arrow(
+    builder: Box<VortexScanBuilder>,
+    out_array: *mut u8,
+    out_schema: *mut u8,
```

Comment:
are these really `*mut u8` or `*mut c_void`?

Reply:
cxx does not understand it. You have to define another struct with the same layout as `FFI_ArrowArray`, and then it can be used as function parameters.
``` rust
#[cxx::bridge(namespace = "vortex::ffi")]
mod ffi {
    // C-compatible structs for Arrow C ABI
    struct CArrowArray {
        length: i64,
        null_count: i64,
        offset: i64,
        n_buffers: i64,
        n_children: i64,
        buffers: usize,
        children: usize,
        dictionary: usize,
        release: usize,
        private_data: usize,
    }
//...
}
```
I've done that before but switched to the current approach